### ISSUE TITLE
Rename setup-elixir gh action to setup-beam and support rebar3 specification 

### ISSUE THREAD

**Issue by starbelly** [erlef/build-and-packaging-wg#20]
# Background 

~~As discussed per #12 we will create a setup-erlang action ala setup-elixir to aid the community in their CI/CD jobs.~~

After a lot of discussion we have decided to not make a new github action, but instead rename setup-elixir to `setup-beam`. The action will be updated to support specifying a rebar3 version for the erlang only case, as well as making elixir optional for the erlang only case.

# Tasks

The following tasks must be completed to resolve this issue. 

- [x] Rename setup-elixir to setup-beam
- [x] Adjust language in README, etc. where appropriate (i.e., substituting elixir with beam where it makes sense)
- [x] Modify it so you can specify rebar3 which would get installed devoid of elixir 
- [x] Modify the action so elixir is optional (if it isn't already this way)
- [x] Review and test with existing erlang projects to ensure we have covered all bases. 
- [x] Publish a new version of the action

# Related agenda items, issues, and pull requests.

- https://github.com/erlef/setup-erlang/pull/3

**Comment by paulo-ferraz-oliveira** [comment]
I'm more than interested seeing this move forward.

I've "worked" with Erlang containers before, as maintained by https://github.com/erlang/docker-erlang-otp (these include a pre-cooked version of `rebar3` also; probably some other stuff).

I've also worked with the Gleam action, that pull from erlang-solutions.com (not all versions available).

I've also implemented GitHub actions in all libraries I use that didn't use it (20+, I guess).

Having a place where management of these updates could be simplified would be of great help.

Some thoughts:
1. where would we pull the images from? If maintained by the EEF, I propose at least all the tagged versions (we'd choose some minimum) be available, and not just some,
2. could `rebar3` be integrated (even if via an option), targeted for a specific version?

**Comment by starbelly** [comment]
> where would we pull the images from? If maintained by the EEF, I propose at least all the tagged versions (we'd choose some minimum) be available, and not just some

we would definitely pull in images from hex so be in sync with setup-elixir  IMO

> could rebar3 be integrated (even if via an option), targeted for a specific version?

I think that makes sense yes. It should default to the latest stable, but you should be able to over ride that. 

> what flags are used to compile the build, since these might sometimes affect the consumer?


Information on that can be found here : https://github.com/hexpm/bob/tree/master/priv/scripts/otp

**Comment by paulo-ferraz-oliveira** [comment]
Bryan, do you think it would be possible to create an empty repo so we can start working on it?

**Comment by starbelly** [comment]
@paulo-ferraz-oliveira I don't see any reason to wait on this. Making one now and will add you to it

**Comment by paulo-ferraz-oliveira** [comment]
Great stuff.

**Comment by tsloughter** [comment]
It can share all its code with setup-elixir, right?

Or do we need to do `setup-beam` to make that possible?

**Comment by paulo-ferraz-oliveira** [comment]
@tsloughter: I don't understand what you mean by "share". Like pulling it in like a dependency?

`setup-beam` would be doable, sure (so that we don't have to copy-paste so much, I guess).

I'm not sure how an action depends on another, but it's a question of looking around, I guess. I'm toying with this already at https://github.com/erlef/setup-erlang/pull/1.

**Comment by paulo-ferraz-oliveira** [comment]
Actually, since the `setup-erlang` action installs Erlang/OTP and (optionally) `rebar3` (which fits into the `setup-elixir` stuff) maybe later the latter can depend on the former, once it's running OK.

**Comment by tsloughter** [comment]
I mean only have 1 repo.

**Comment by paulo-ferraz-oliveira** [comment]
Ah, that's OK too, I guess. (probably shouldn't have jumped into the implementation, it seems 😕)

**Comment by tsloughter** [comment]
Yea, I just think if we don't do a single repo it'll get overly complicated keeping stuff in sync. And with a single repo/action we can have one place for people to get an action for erlang, elixir, lfe, gleam, etc

**Comment by starbelly** [comment]
I like the idea @tsloughter. @paulo-ferraz-oliveira we haven't published anything, so it's early on enough to rename, etc.

**Comment by wojtekmach** [comment]
Re renaming, given GitHub creates redirects after renaming the repository, I expect GitHub Actions won't have a problem with that at all. And so if the plan is to only have `erlef/setup-beam` project, I'd plan to eventually rename `erlef/setup-elixir` to _that_.

I'd start with a fork and when it's ready, merge the changes back in, and rename to `erlef/setup-beam`. I'd expect most complexity would be around installing OTP itself (it's OS & architecture dependant) and installing all the other languages should be hopefully fairly straightforward. (And if it isn't, it means we need to improve tooling first.)

**Comment by starbelly** [comment]
@wojtekmach I'm fine with that approach, no strong feelings either way. Which ever way is less "dangerous" and the least confusing for users.

**Comment by starbelly** [comment]
@tsloughter to be clear, it _could_ get overly complicated, that is a valid concern, but I'm also not sure it would either. I think if we started to approach that threshold we could break out. Thoughts?

**Comment by paulo-ferraz-oliveira** [comment]
Shall I drop my pull request then?

**Comment by paulo-ferraz-oliveira** [comment]
Keeping things separate is probably not that problematic. Are we worried that we'll duplicate much code, is that it?

Keeping this separate avoids having to deploy a new version when only `elixir` OR `erlang` needs an update, in one of the two. And I believe that reusability can be achieved by having one action depend on the other, not necessarily by mixing the two, pretty much like Elixir builds on top of the VM 😄.

**Comment by wojtekmach** [comment]
then by all means explore the separate projects route!

**Comment by paulo-ferraz-oliveira** [comment]
In any case, I'll keep my eyes open, and continue with `setup-erlang`. It's a good learning experience as well as a new set of eyes on `setup-elixir`.

**Comment by paulo-ferraz-oliveira** [comment]
@wojtekmach, I'm proposing, not deciding 😄 Whatever route is decided, it'll surely best serve everybody. (and it's not that much code that we can't easily merge in the future) - took me a couple hours to "copy-paste and adapt".

**Comment by starbelly** [comment]
I'm fine with proceeding. I think we may not have enough cases yet perhaps. I think if we want one action, we can have that. If we want modularization, we can change it, and it wouldn't affect the user either.

Let's keep this discussion going on though, it's great.

**Comment by wojtekmach** [comment]
Unless someone has a strong opinion, I think it's fine if whoever ends up working on it, picks whatever approach makes sense to them. And when we have a 3rd one (after setup-erlang & setup-elixir) we can check in what's the best way to support that. (Rule of three [1] comes to mind.)

[1] https://en.wikipedia.org/wiki/Rule_of_three_(computer_programming)

**Comment by starbelly** [comment]
@wojtekmach 💯 I literally just stated that on slack. Specifically, in general when I write code, I don't try to generalize until I have at **least** three cases.  Also, TIL about this rule, now I've got something to point to. Cheers 👍

**Comment by tsloughter** [comment]
I partly brought it up since I already use `setup-elixir` in some places where I only need Erlang. So I didn't think there was a reason to have a new and separate project.

**Comment by wojtekmach** [comment]
Yup, same here.

If we'd go with one project, I think the way it would work is `otp-version` would be the only required input and `elixir-version`, `lfe-version` etc would default to `null` (just like here: https://github.com/erlef/setup-elixir/blob/885971a72ed1f9240973bd92ab57af8c1aa68f24/action.yml#L17) which means that thing isn't installed.

**Comment by paulo-ferraz-oliveira** [comment]
Agreed. As discussed with @starbelly offline, I will proceed with this one project and see where it goes. If tomorrow a third element comes that copies heavily from one of these actions we might better think about generalising it, which hopefully won't be too hard. It's just a couple JavaScript files and a shell script, mostly.

**Comment by starbelly** [comment]
@tsloughter The thing is, I see and agree with you on that as well and I really fancy the idea of having **one**.  We have a growing problem (the way I see it) that you had alluded to on erlef slack one time. Specifically, instead of everyone focusing on one solution, we have to have N initiatives to get N solutions in place. I see this as similar to Linux distributions in the early days and what a pain that was (and perhaps still is).  I'd like if we can put more effort into making such problems simply go away. 

I also am head over heels with the idea of being inclusive in how we refer to all languages (i.e., setup-beam). Even if it's for the sake of not having to constantly repeat a list. 

I believe there will be one, indeed, but it seems like things aren't clear yet to some. I honestly though just don't have strong feelings on this at the moment one way or the other. I'd like there to be one, but if it's not clear how this will end up, I also support doing a few cases first.

**Comment by paulo-ferraz-oliveira** [comment]
> Specifically, instead of everyone focusing on one solution, we have to have N initiatives to get N solutions in place

I'd spoken about this to you earlier, also, @starbelly. The fact that I feel some efforts need to go in the same direction, but here we're not preventing that from happening (I hope!). As I stated before, probably what we're doing will end up merged with whatever `setup-elixir` did for the Erlang/OTP part in a different action (at least half of what I committed was copy-paste, anyway).

**Comment by starbelly** [comment]
I do think we will end up renaming `setup-elixir`. I think the issues that @paulo-ferraz-oliveira opened up on setup-elixir highlights a big reason why. Even if we have a base action, it'll be no fun to deal with little small issues in javascript etc across multiple repos. Having one repo vs N simplifies for the community as well.  Further, per the nature of what we're dealing with,  I don't think it will get overly-complicated either, the cases are finite if you will, and even if it does, you just fork and create a new action. However, I think there will be many great changes that will fall out of what @paulo-ferraz-oliveira is doing on the setup-erlang repo, but ultimately it will all collapse into one.

**Comment by wojtekmach** [comment]
@paulo-ferraz-oliveira you may know this already but recently I found two interesting things about GitHub Actions:

1. `composite` action type which is basically an alternative to writing JS. **It allows re-using existing actions too.** See https://docs.github.com/en/actions/creating-actions/creating-a-composite-run-steps-action#creating-an-action-metadata-file
2. https://github.com/nektos/act - allows testing actions locally. There's a lot of missing stuff (notably supporting `echo ... >> $GITHUB_PATH` that a lot of actions rely on) but oh so much better than committing and waiting for CI result. (There's an open PR to support `composite` action: https://github.com/nektos/act/pull/514, worked great for me.)

might be helpful when developing actions, was certainly useful to me.

**Comment by wojtekmach** [comment]
Btw, I just created (_another_) proof-of-concept: https://github.com/wojtekmach/beamup. There's a ton of work to get it past the finish line so take this with a grain of salt. I really wanted to make it as self-contained as possible but if there are any good ideas here, might make sense to extract them out as separate tools. Again, just an experiment but hopefully it can inspire folks to experiment so more. :)

**Comment by tsloughter** [comment]
Oh, is beamup for CI? As opposed to erlup (https://github.com/tsloughter/erlup/) -- which I want to eventually use the binary releases and also support other beam languages. So if you do move to being for only CI then it'd be great to free up the `beamup` name for a local dev tool like the other `*up` tools so `erlup` can become `beamup` :)

**Comment by starbelly** [comment]
Yeah, I didn't think beamup was for ci, but if it might be, this would obsolete bob, no?

**Comment by paulo-ferraz-oliveira** [comment]
> composite action type which is basically an alternative to writing JS.

Didn't know that, thanks. I followed GitHub actions tutorials and looked at `setup-elixir` alone.

> It allows re-using existing actions too.

This shall be interesting.

Thanks for the info.

**Comment by wojtekmach** [comment]
Yeah sorry if its misleading, the project is called beamup, which has specific connotations, but its a collection of things, *up being just one of them.

 I have the installer part (./install.sh), an equivalent to erlup and friends for dev, and the GH action (./setup-beam) is just a tiny wrapper on top. so while these two tools have different use cases and audiences (dev vs ci), i think the proof of concept nicely shows that one can depend on another.

agreed a single beamup that can install binary releases for different languages would be nice (thats exactly what this install.sh piece already does :))

**Comment by wojtekmach** [comment]
Yes, we are planning to retire pieces of Bob in favour of using GH actions.

**Comment by starbelly** [comment]
Definitely good to know :)

**Comment by starbelly** [comment]
@wojtekmach Can you expand more on how you think this might work in practice? 

Some things I can think of :

- You have a repository that uses `build-otp` or something similar
- The action gets triggered when otp, elixir, etc. creates a new tag 
- The action builds otp, elixir, etc. and stores the asset as a release on github 

Is this your thinking?

**Comment by wojtekmach** [comment]
Happy to elaborate but not sure what you are asking, could you clarify?

Are you asking if there is one GitHub repo that polls for new OTP, Elixir, Gleam, etc release events and based on that builds a binary and uploads it back to the release? If so, imho no, I think this should be decentralised and each project (OTP, Elixir, Gleam, etc) should continue to build its own release assets like it does today and the reason is moving it elsewhere would make these projects harder to maintain, you don't want to change how Elixir is built (change its Makefile) in _another_ repo, that wouldn't be good.

FWIW I did create a separate repo (wojtekmach/beamup) that builds OTP (and in fact yet another, wojtekmach/otp_releases that holds releases) rather than doing it in erlang/otp fork and in hindsight I should have done it in a fork.

**Comment by starbelly** [comment]
> FWIW I did create a separate repo (wojtekmach/beamup) that builds OTP (and in fact yet another, wojtekmach/otp_releases that holds releases) rather than doing it in erlang/otp fork and in hindsight I should have done it in a fork.

That's why I thought you might have been suggesting something similar to what I was talking about, but not for general distribution, for action specific purposes. But yeah, I agree, I was about to say I believe I remember Jon talking about providing more builds for different platforms.

**Comment by starbelly** [comment]
I personally have made up my mind per @wojtekmach's suggestion around using composite actions which negates having to fool with js for a single repository, be it setup-beam or setup-otp. Even with the js, it really isn't much at all, I re-realized that after reviewing setup-elixir and the current PR to setup-erlang.  I think we should incorporate changes made by @paulo-ferraz-oliveira that are appropriate to apply in said repo. I also think we should simply rename setup-elixir, as @wojtekmach pointed out we just got a bunch of people to switch over to erlef/setup-elixir and he also tested to check that a rename won't break an action, thus it is safe. 

I'd like @ericmj to be aware and add thoughts if he has them or simply state that he agrees since he's the other maintainer on setup-elixir and got the ball rolling.

The majority is in favor of setup-beam, which includes myself, however it should be noted some have raised concerns around it not being technically correct and favor setup-otp.  So, also need to hear from @ericmj as to what he thinks.

**Comment by wojtekmach** [comment]
> I personally have made up my mind per @wojtekmach's suggestion around using composite actions which negates having to fool with js for a single repository, be it setup-beam or setup-otp. 

Just to be clear, I think that if we would to write it from scratch, I'd definitely use the `composite` approach, but given setup-elixir is already widely used and written in JS, it's probably more pragmatic to leave it as is and spend time elsewhere. I haven't contributed to it all, I don't know how easy or hard it is to maintain it in JS. (That being said, if someone does want to rewrite, they have my vote.)

If we go with a single setup-* project based on setup-elixir then I believe the only thing we need to do is making `elixir-version` input argument optional, isn't it?

Regarding how to actually call the single project, the obvious choices are setup-beam/setup-otp/setup-erlang. I think I've come around to preferring setup-erlang actually. I think that might be the most accessible to newcomers, they probably heard about Erlang, they are probably unclear what OTP exactly is, and they probably haven't heard about BEAM yet. But maybe it would be counterintuitive that `setup-erlang` can be used to install Elixir (and maybe Gleam and others in the future)? I don't have a strong opinion about naming.

**Comment by starbelly** [comment]
Yeah, I basically see a re-write happening if you will, where a fork or branch is made, composite action is implemented, scripts are adjusted as need be, etc. Merged into main, tagged and published as v2. 

> If we go with a single setup-* project based on setup-elixir then I believe the only thing we need to do is making elixir-version input argument optional, rename setup-elixir to setup-<x> and that's it, isn't it?

Correct, to start with, then add support for install rebar3, optionally, and that's how I would kick it off myself. 

> But maybe it would be counterintuitive that setup-erlang

yes, that's part of my thinking, it is counter-intuitive. If it handles quite a bit more than erlang itself, then I don't think it makes sense.

**Comment by wojtekmach** [comment]
Unless theres's a breaking change, there's no reason to publish v2. If
there's v2, people would have to update their workflow files and they
just did that recently when switching from ***@***.*** to
***@***.***

On March 15, 2021, GitHub ***@***.***> wrote:
> Yeah, I basically see a re-write happening if you will, where a fork
> or branch is made, composite action is implemented, scripts are
> adjusted as need be, etc. Merged into main, tagged and published as
> v2.
>
> > If we go with a single setup-* project based on setup-elixir then I
> > believe the only thing we need to do is making elixir-version input
> > argument optional, rename setup-elixir to setup- and that's it,
> > isn't it?
> >
>
> Correct, to start with, then add support for install rebar3,
> optionally, and that's how I would kick it off myself.
>
> > But maybe it would be counterintuitive that setup-erlang
> >
>
> yes, that's part of my thinking, it is counter-intuitive. If it
> handles quite a bit more than erlang itself, then I don't think it
> makes sense.
>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/erlef/build-and-packaging-wg/issues/20#issuecomment-
799523331>, or unsubscribe
<https://github.com/notifications/unsubscribe-
auth/AAASSJ3L7YC2V7ZQ3Z6OOSTTDYTLHANCNFSM4YVSUVDA>.

**Comment by starbelly** [comment]
If we are sure it would not break anything, then sure. I'm fine with that.

**Comment by starbelly** [comment]
Ok. So after discussing some things with @ericmj we have decided the following : 

- we will not jump ship from js to composite actions until there is a very simple solution around handling inexact versions (appreciate @paulo-ferraz-oliveira's thoughts on this). I personally don't have time for this at the moment. 
- We will rename setup-elixir to setup-beam, Eric and I prefer this, and the majority of the group and other interested parties sided with setup-beam. 
- We would like for @paulo-ferraz-oliveira to incorporate the improvements he made into setup-beam if he wishes to do that
- We will only support rebar3 specification and making elixir optional (if it isn't already) out of the gate. We need to build up to other languages, but it's not something that has to be done right now.

This was a lot of discussion, back and forth, changing of minds, etc. but I feel it was worth it and I vastly appreciate everyone's thoughts and input. ❤️

**Comment by paulo-ferraz-oliveira** [comment]
Yeah, I'll try to find sometime for this. I knew we'd find a proper first solution that pleased more than "just Erlang" or "just Elixir". Thanks.

**Comment by paulo-ferraz-oliveira** [comment]
https://github.com/erlef/setup-elixir/pull/9: all green. Review ahoy.

**Comment by paulo-ferraz-oliveira** [comment]
@starbelly, do you still think this needs to stay open, given the recent developments in `setup-beam`? Only action that's not done (from the initial description) is "in-the-wild tests".

**Comment by starbelly** [comment]
@paulo-ferraz-oliveira Hmm, it's that and the last one, the release. Which we will do on monday. As stated, we can do a tag today, but we'll do the release on Monday. Then this can be closed. If we closed it now I'd be breaking protocol :)

**Comment by paulo-ferraz-oliveira** [comment]
I updated `package.json` and pushed `v1.7.0` with it. I see some `vx.y` companions to `vx.y.z` but it's somewhat inconsistent. I chose not to push `v1.7` (there's `v1.5` for example, but no `v1.6`).

| "branch" | tags
|- |-
| 1.0 | v1.0.0/1.0.0
| 1.1 | v1.1.0
| 1.2 | v1.2.0
| 1.2 | v1.2.1, v1.2.2
| 1.3 | v1.3.0/v1.3
| 1.4 | v1.4.0, v1.4.1/v1.4, v1.4.2
| 1.5 | v1.5.0/v1.5
| 1.6 | v1.6.0, v1
| 1.7 | v1.7.0

By my calculations:
* `1.0.0` is inconsistent with all the others (no `v`)
* `v1.4` should be paired with `v1.4.2`, not `v1.4.1`
* `v1.6` could potentially exist for `v1.6.0` (the same for "branch" 1.7 - unless we assume we're not doing that anymore)

**Comment by starbelly** [comment]
@paulo-ferraz-oliveira I believe the rationale here is this allows someone to say `@v1`, `@v1.7`, or `@v1.7.3`. I think the inconsistency is probably related to changes in how GitHub actions work now vs how they worked when first rolled out.

**Comment by paulo-ferraz-oliveira** [comment]
I understand there may be a rationale behind it 😄, nonetheless, I'm still confused with:
* `1.0.0` not having `v`
* `v1.4.1` being paired with `v1.4`, and not `v1.4.2`

Should we push `v1.6` and `v1.7` where applicable?

**Comment by starbelly** [comment]
@paulo-ferraz-oliveira I think so yes

**Comment by paulo-ferraz-oliveira** [comment]
`v1.6` and `v1.7` pushed. Apart from the `1.0.0` idiosyncrasy I still thing we should maybe move `v1.4` to `v1.4.2`.
